File: gene/python3/gene.py
import random

class Gene:
    GENE_COUNT = 18
    CORRECT_GENE = ['c', 'a', 'r', 'a', 'm', 'e', 'l', 'f', 'r', 'a', 'p', 'p', 'u', 'c', 'h', 'i', 'n', 'o']
    ALPHABET = [
            'a', 'b', 'c', 'd', 'e',
            'f', 'g', 'h', 'i', 'j',
            'k', 'l', 'm', 'n', 'o',
            'p', 'q', 'r', 's', 't',
            'u', 'v', 'w', 'x', 'y', 'z'
            ]
    def __init__(self):
        self.__gene = []
        self.__point = 0
        for _ in range(self.GENE_COUNT):
            self.__gene.append(random.choice(self.ALPHABET))

    def get_gene(self): return self.__gene
    def get_gene_index(self, index): return self.__gene[index]
    def set_gene_index(self, index, value): self.__gene[index] = value
    def get_point(self): return self.__point
    def set_point(self, point): self.__point = point

    def review(self):
        self.__point = 0
        i = 0
        for g in self.__gene:
            if g == self.CORRECT_GENE[i]: self.__point += 1
            i += 1

    def mutation(self):
        num1 = random.randint(0, self.GENE_COUNT - 1)
        num2 = random.randint(0, self.GENE_COUNT - 1)
        self.__gene[num1] = random.choice(self.ALPHABET)
        self.__gene[num2] = random.choice(self.ALPHABET)

    @staticmethod
    def make_child(father, mother):
        rnd = random.randint(0, Gene.GENE_COUNT - 1)
        child = Gene()
        for i in range(Gene.GENE_COUNT):
            if i > rnd:
                mother_g = mother.get_gene_index(i)
                child.set_gene_index(i, mother_g)
            else:
                father_g = father.get_gene_index(i)
                child.set_gene_index(i, father_g)
        return child


class Colony:
    GENERATION = 20
    DESTROY = 5

    def __init__(self):
        self.__colony = []
        for _ in range(self.GENERATION ):
            self.__colony.append(Gene())

    def get_colony(self): return self.__colony;

    def reviews(self):
        for c in self.__colony:
            c.review()

    def delete(self):
        for _ in range(self.DESTROY):
            s_key = 0
            s_point = 0
            i = 0
            for c in self.__colony:
                point = c.get_point()
                if i == 0:
                    s_point = point
                elif s_point > point:
                    s_key = i
                    s_point = point
                i += 1
        pass

    def make_children(self):
        for _ in range(len(self.__colony)):
            father = random.randint(0, self.GENERATION - self.DESTROY - 1)
            mother = random.randint(0, self.GENERATION - self.DESTROY - 1)
            child = Gene.make_child(self.__colony[father], self.__colony[mother])
            self.__colony.append(child)

    def mutations(self):
        for c in self.__colony:
            rnd = random.randint(0, 25)
            if rnd == 1:
                c.mutation()

def main():
    generations = Colony()
    i = 1
    while True:
        generations.reviews()

        point = 0
        genes_s = ""
        for c in generations.get_colony():
            tmp_s = ""
            for g in c.get_gene():
                tmp_s += g

            point = c.get_point()
            gene_s = tmp_s
            print(gene_s, point, "generations:", i)
            if point >= Gene.GENE_COUNT:
                print("finish")
                break

            generations.delete()
            generations.make_children()
            generations.mutations()
            i += 1

if __name__ == '__main__':
    main()

File: gene/ruby/gene.rb
class Gene
      GENE_COUNT = 18
        #Caramel Frappucchino
      CORRECT_GENE = ['c', 'a', 'r', 'a', 'm', 'e', 'l', 'f', 'r', 'a', 'p', 'p', 'u', 'c', 'h', 'i', 'n', 'o']
      ALPHABET = [
            'a', 'b', 'c', 'd', 'e',
            'f', 'g', 'h', 'i', 'j',
            'k', 'l', 'm', 'n', 'o',
            'p', 'q', 'r', 's', 't',
            'u', 'v', 'w', 'x', 'y', 'z'
      ]
      def initialize()
            @gene = []
            @point = 0
            GENE_COUNT.times do
                  @gene.push(ALPHABET.sample())
            end
      end

      #評価
      def review()
            @point = 0
            @gene.each_with_index do |g, i|
                  if g == CORRECT_GENE[i]
                        @point += 1
                  end

            end
      end

      #突然変異
      def mutation()
            num1 = rand(0..GENE_COUNT)
            num2 = rand(0..GENE_COUNT)
            @gene[num1] = ALPHABET.sample()
            @gene[num2] = ALPHABET.sample()
      end

      #交配
      def self.make_child(mother, father)
            rnd = rand(0..GENE_COUNT)
            child = Gene.new()
            GENE_COUNT.times do |i|
                  if i < rnd
                        m_gene = mother.gene[i]
                        child.gene[i] = mother.gene[i]
                  else
                        f_gene = father.gene[i]
                        child.gene[i] = father.gene[i]
                  end
            end
            return child
      end
      attr_accessor :gene, :point
end


class Colony
      GENERATIONS = 20 #1世代の遺伝子数
      DESTROY = 5 #1世代の淘汰数

      def initialize()
            @colony = []
            GENERATIONS.times do
                  @colony.push(Gene.new())
            end
      end

      #評価
      def reviews()
            @colony.each do |c|
                  c.review()
            end
      end

      #淘汰
      def delete()
            DESTROY.times do
                  s_key = 0
                  s_point = 0
                  @colony.each_with_index do |c, i|
                        if i == 0
                              s_point = c.point
                        elsif s_point > c.point
                              s_key = i
                              s_point = c.point
                        end
                  end
                  @colony.delete_at(s_key)
            end
      end

      #交配
      def make_children()
            DESTROY.times do
                  father = rand(0..GENERATIONS - DESTROY - 1)
                  mother = rand(0..GENERATIONS - DESTROY - 1)
                  child = Gene::make_child(@colony[father], @colony[mother])
                  @colony.push(child)
            end
      end

      def mutations()
            @colony.each do |c|
                  rnd = rand(0..25)
                  if rnd == 1
                        c.mutation()
                  end
            end
      end

      attr_accessor :colony
end


##main処理

generations = Colony.new()

i = 1
while true
      generations.reviews()

      point = 0
      gene_s = ""
      generations.colony.each_with_index do |c, i|
            tmp_s = ""
            c.gene.each do |g|
                  tmp_s += g
            end
            if i == 0
                  point = c.point
                  gene_s = tmp_s
            else
                  point = c.point
                  gene_s = tmp_s
            end
      end
      puts "#{gene_s}: #{point}/ generation: #{i}"
      if point >= Gene::GENE_COUNT
            puts "finish"
            break
      end

      generations.delete()
      generations.make_children()
      generations.mutations()
#      sleep(0.01)
      i += 1
end
